File: src/rust/src/combination.rs
use crate::{
    factorial_table::factorial_table,
    inverse_factorial_table::inverse_factorial_table,
    ops::MulInv,
};

pub struct Combination<T> {
    fact: Vec<T>,
    inv_fact: Vec<T>,
}

impl<T> Combination<T>
where
    T: std::ops::Mul<Output = T> + From<u64> + Clone,
{
    pub fn new(size: usize) -> Self
    where
        T: MulInv<Output = T>,
    {
        Self {
            fact: factorial_table::<T>(size),
            inv_fact: inverse_factorial_table::<T>(size),
        }
    }

    pub fn calc(&self, n: usize, k: usize) -> T {
        if k > n {
            return 0.into();
        }
        self.fact[n].clone()
            * self.inv_fact[n - k].clone()
            * self.inv_fact[k].clone()
    }

    pub fn inv(&self, n: usize, k: usize) -> T {
        assert!(k <= n); // (n, k) := 0 if k > n, so the inverse is undefined.
        self.inv_fact[n].clone()
            * self.fact[k].clone()
            * self.fact[n - k].clone()
    }
}

// #[cfg(test)]
mod tests {

    #[test]
    fn test() {
        use super::*;
        use crate::modular::{
            arithmetic::Modular1_000_000_007,
            int::Modint as StaticModularInt,
        };
        type Mint = StaticModularInt<u32, Modular1_000_000_007>;
        let choose = Combination::<Mint>::new(100);
        assert_eq!(choose.calc(5, 2), 10.into());
    }
}

File: src/python/src/dsalgo/combination.py
import typing
from dsalgo.factorial_table import factorial_table
from dsalgo.inverse_factorial_table import inverse_factorial_table

import unittest


def combination(
    mul: typing.Callable[[int, int], int],
    inv: typing.Callable[[int], int],
    size: int,
) -> typing.Callable[[int, int], int]:
    fact = factorial_table(mul, size)
    ifact = inverse_factorial_table(mul, inv, size)

    def f(n: int, k: int) -> int:
        if not 0 <= k <= n:
            return 0
        return mul(mul(fact[n], ifact[n - k]), ifact[k])

    return f


class Tests(unittest.TestCase):
    def test(self) -> None:
        mod = 1_000_000_007
        f = combination(
            lambda x, y: x * y % mod, lambda x: pow(x, mod - 2, mod), 100
        )
        assert f(10, 2) == 45


if __name__ == "__main__":
    import doctest

    unittest.main()

    doctest.testmod(verbose=True)
